internal/handlers/commonpath/commonpath.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package commonpath contains functions used for path.
+package commonpath
+
+import (
+	"errors"
+	"strconv"
+
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+)
+
+// FindValuesOpts sets options for FindValues.
+type FindValuesOpts struct {
+	// SearchInArray searches by iterating through the whole array to find documents that contains path key.
+	// Using path `v.foo` and `v` is an array, it returns all document which has key `foo`.
+	// If `v` is not an array, SearchInArray has no impact.
+	SearchInArray bool
+	// FindArrayIndex gets an element at the specified index of an array.
+	// Using path `v.0` and `v` is an array, it returns 0-th index element of the array.
+	// If `v` is not an array, FindArrayIndex has no impact.
+	FindArrayIndex bool
+}
+
+// FindValues goes through each key of the path iteratively on doc to find values
+// at the suffix of the path. At each key of the path, it checks:
+//   - if the document has the key;
+//   - if the array contains an element at an index where index equals to the key, and;
+//   - if the array contains one or more documents which have the key.
+//
+// It returns a slice of values and an empty array is returned if no value was found.
```

Comment:
I think it is extremely hard to understand what this function does due to confusing comments and inconsistent terminology. Let's read it together:

> FindValues goes through each key of the path iteratively on doc

It is not clear if we iterate over path elements (that could contain array indexes, not only document keys) or over doc fields.

> to find values at the suffix of the path.

?.. If we iterate over path elements, then what is the path suffix if we already iterate over path elements? If we iterate over doc fields, then why do we use the path's [suffix](https://github.com/FerretDB/FerretDB/blob/19e1cdbfde91cafad281d7b18e02c9dbf3a8b233/internal/types/path.go#L134) instead of just passing the single element?

> At each key of the path

Ok, at each element of the path. So we iterate over path elements after all.

> it checks if the document has the key;

But what if that's not a document or not a key?

> if the array contains an element at an index where index equals to the key

Ok, so "element at an index" is a value at the given index, I guess. Why do we compare the value with the key?..

> if the array contains one or more documents which have the key.

Ok, so we got a boolean value, true or false. What do we do with it?

> It returns a slice of values and an empty array is returned if no value was found.

I'm completely lost.

---

I understand that the whole area is complicated, but that's a reason to make it approachable. If we keep it as is, very soon, that will be magic no one would understand and touch. Personally, after spending days (and hours of wall clock time) on it, my head hurts – and I still don't get it. (If anything, I get less after today 😬)

I think we could try to simplify it by:

1. Using consistent terminology. I think it is already defined by the type package; we could also improve it if needed.
2. Splitting that function. Right now, the cyclomatic complexity is very high. If we split it into multiple functions with good names and descriptions, someone could get a better idea of what the whole thing does by reading the code.
3. Writing better tests (see below). Failing to understand description and code, someone at least could read the test.

Reply:
Let me rework on this. Reading your comment, I think I'm also getting lost which is really bad sign. Splitting to functions would improve it too.

Reply:
Splitted them to smaller functions and updated doc string to follow closer to path package. Let me know if this improves understanding. or how we can make it better :pray: